haystack/preview/document_stores/memory/document_store.py

```diff
@@ -259,3 +271,104 @@ def bm25_retrieval(
+
+        :param query_embedding: Embedding of the query.
+        :param filters: A dictionary with filters to narrow down the search space.
+        :param top_k: The number of top documents to retrieve. Default is 10.
+        :param scale_score: Whether to scale the scores of the retrieved Documents. Default is True.
+        :param return_embedding: Whether to return the embedding of the retrieved Documents. Default is False.
+        :return: A list of the top 'k' documents most relevant to the query.
+        """
+        if len(query_embedding) == 0 or not isinstance(query_embedding[0], float):
+            raise ValueError("query_embedding should be a non-empty list of floats.")
+
+        filters = filters or {}
+        all_documents = self.filter_documents(filters=filters)
+
+        documents_with_embeddings = [doc for doc in all_documents if doc.embedding is not None]
+        if len(documents_with_embeddings) == 0:
+            logger.info(
+                "No Documents found with embeddings. Returning empty list. "
+                "To generate embeddings, use a DocumentEmbedder."
+            )
+            return []
+        elif len(documents_with_embeddings) < len(all_documents):
+            logger.info(
+                "Skipping some Documents that don't have an embedding. "
+                "To generate embeddings, use a DocumentEmbedder."
+            )
+
+        # mypy complains that doc.embedding could be None, but we already excluded all Documents with None embedding
+        embedding_sizes = [len(doc.embedding) for doc in documents_with_embeddings]  # type: ignore[arg-type]
+        if any(size != embedding_sizes[0] for size in embedding_sizes):
+            raise ValueError(
+                "The embedding size of all Documents should be the same."
+                "Please make sure that the Documents have been embedded with the same model."
+            )
+
+        if len(query_embedding) != embedding_sizes[0]:
+            raise ValueError(
+                "The embedding size of the query should be the same as the embedding size of the Documents."
+                "Please make sure that the query has been embedded with the same model as the Documents."
+            )
+
+        scores = self._compute_query_embedding_similarity_scores(
+            embedding=query_embedding, documents=documents_with_embeddings, scale_score=scale_score
+        )
+
+        # create Documents with the similarity score for the top k results
+        top_documents = []
+        for doc, score in sorted(zip(documents_with_embeddings, scores), key=lambda x: x[1], reverse=True)[:top_k]:
+            doc_fields = doc.to_dict()
+            doc_fields["score"] = score
+            if return_embedding is False:
+                doc_fields["embedding"] = None
+            top_documents.append(Document(**doc_fields))
+
+        return top_documents
+
+    def _compute_query_embedding_similarity_scores(
```

Comment:
I'm not sure why this is re-implemented where we already have a version that can compute embeddings both on the GPU (with torch) and CPU. See this PR that brought a big speed up to the InMemoryDocumentStore: https://github.com/deepset-ai/haystack/pull/2091

I don't know how this variant compares though. In general, I think for this type of tensor operation code it is important to run some perf benchmarks as small changes can bring a lot of benefit.

Reply:
Hello @mathislucka, thanks for raising this point, which gave an opportunity for an internal discussion.

1) The `MemoryDocumentStore` is a simple Document Store, which should have very few dependencies and is not intended for production.
2) If feasible (not very slow), we would prefer that the `MemoryDocumentStore` use only `numpy` and not `pytorch`, which is a heavy dependency.
3) [#2091](https://github.com/deepset-ai/haystack/pull/2091) introduced nice performance improvements but was a bit buggy (see [#4113](https://github.com/deepset-ai/haystack/pull/4113)), so I am taking a cue from that implementation (the `numpy` part) and trying to improve it.

Reply:
If you want to run experiments on an EC2 machine or do some batch processing then the InMemory document store can be an excellent option if it's fast. Probably all other doc stores would introduce some overhead for network calls and such. I also liked to have the InMemory store to just do some local embedding distance calculations without needing to reach for a separate library. It's not about production, it's about using the inmemory document store in your data science workflows. Of course, that can be deprioritized if you feel like it's more important to keep the dependency footprint small. 

How will people calculate embeddings without torch? Just use OpenAI?

Reply:
At the moment, we aim to minimize dependencies for the `MemoryDocumentStore`.
 
If `numpy` implementation is impractical/too slow, we will add GPU support in a separate PR.